File: src/components/TramTracker.js
import * as THREE from 'three';

class TramTracker {
  constructor() {
    // Simplified tram tracking state - mainly for frontend display
    this.currentStatus = 'Stopped';
    this.currentLocation = null;
    this.lastLocation = null;
    this.isMoving = false;
    this.nextBuilding = null;
    this.lastPassedBuilding = null;
    
    // Movement detection settings
    this.movementThreshold = 0.00001; // GPS coordinate difference to detect movement
    this.stoppedDuration = 30 * 60 * 1000; // 30 minutes in milliseconds
    
    // Tracking history
    this.locationHistory = [];
    this.maxHistoryLength = 10;
    this.lastUpdateTime = Date.now();
    this.lastMovementTime = Date.now();
    
    // Updated building coordinates with optimized radius to match backend
    this.buildings = [
      {
        id: 'msm_building',
        name: 'MSM Building',
        displayName: 'MSM Building',
        lat: 13.612565,
        lon: 100.836516,
        radius: 0.0005,  // Optimized radius: ~55 meters for reliable detection
        stopIndex: 0
      },
      {
        id: 'it_building',
        name: 'IT Building',
        displayName: 'IT Building',
        lat: 13.612177,
        lon: 100.836425,
        radius: 0.0005,  // Optimized radius: ~55 meters for reliable detection
        stopIndex: 1
      },
      {
        id: 'au_mall',
        name: 'AU Mall',
        displayName: 'AU Mall',
        lat: 13.612764,
        lon: 100.833440,
        radius: 0.0005,  // Optimized radius: ~55 meters for reliable detection
        stopIndex: 2
      },
      {
        id: 'queen_of_sheba',
        name: 'Queen of Sheba',
        displayName: 'Queen of Sheba',
        lat: 13.614219,
        lon: 100.832132,
        radius: 0.0005,  // Optimized radius: ~55 meters for reliable detection
        stopIndex: 3
      }
    ];
    
    // Continuous loop sequence - MSM → IT → AU Mall → Queen of Sheba → MSM
    this.loopSequence = [
      'msm_building',
      'it_building',
      'au_mall',
      'queen_of_sheba'
    ];
    
    // Status change callbacks
    this.statusChangeCallbacks = [];
    this.lastNotifiedStatus = null;
    
    console.log('🚋 TramTracker initialized with continuous loop logic');
    console.log('📍 Tram stops (continuous loop):', this.buildings.map(b => b.displayName).join(' → ') + ' → MSM');
  }
  
  // Get next building in continuous loop: MSM → IT → AU Mall → Queen of Sheba → MSM
  getNextBuilding(currentBuildingId) {
    if (!currentBuildingId) {
      return this.buildings[0]; // Start with MSM
    }
    
    try {
      // Find current building in the loop sequence
      const currentIndex = this.loopSequence.findIndex(id => id === currentBuildingId);
      if (currentIndex === -1) {
        return this.buildings[0]; // Default to MSM if not found
      }
      
      // Get next building in loop (wraps around to MSM after Queen of Sheba)
      const nextIndex = (currentIndex + 1) % this.loopSequence.length;
      const nextBuildingId = this.loopSequence[nextIndex];
      
      // Find the building object
      return this.buildings.find(b => b.id === nextBuildingId) || this.buildings[0];
      
    } catch (error) {
      console.error('Error getting next building:', error);
      return this.buildings[0]; // Default to MSM
    }
  }
  
  // Improved position update with continuous loop logic
  updatePosition(lat, lon) {
    const currentTime = Date.now();
    const newLocation = { lat, lon, timestamp: currentTime };
    
    // Store previous location
    this.lastLocation = this.currentLocation;
    this.currentLocation = newLocation;
    
    // Add to history
    this.addToHistory(newLocation);
    
    // Detect movement
    this.detectMovement();
    
    // Check for building detection
    this.detectBuilding(lat, lon);
    
    // Update status with continuous loop logic
    this.updateStatus();
    
    this.lastUpdateTime = currentTime;
    
    return this.getTrackingInfo();
  }
  
  addToHistory(location) {
    this.locationHistory.push(location);
    if (this.locationHistory.length > this.maxHistoryLength) {
      this.locationHistory.shift();
    }
  }
  
  // Detect if tram is near a building
  detectBuilding(lat, lon) {
    for (const building of this.buildings) {
      const distance = this.calculateDistance(lat, lon, building.lat, building.lon);
      
      if (distance <= building.radius) {
        // We're inside a building's radius
        if (!this.lastPassedBuilding || building.id !== this.lastPassedBuilding.id) {
          // New building detected
          console.log(`🏢 Frontend: Tram entered building: ${building.displayName}`);
          this.lastPassedBuilding = building;
          this.nextBuilding = this.getNextBuilding(building.id);
          console.log(`🎯 Frontend: Next destination updated: ${this.nextBuilding.displayName} (continuous loop)`);
        }
        return building;
      }
    }
    return null;
  }
  
  // Improved movement detection
  detectMovement() {
    if (!this.lastLocation || !this.currentLocation) {
      this.isMoving = false;
      return;
    }
    
    // Calculate distance moved
    const distance = this.calculateDistance(
      this.lastLocation.lat, this.lastLocation.lon,
      this.currentLocation.lat, this.currentLocation.lon
    );
    
    // Check if moved significantly (more sensitive threshold)
    if (distance > this.movementThreshold) {
      this.isMoving = true;
      this.lastMovementTime = Date.now();
    } else {
      // Check if stopped for 30 minutes
      const timeSinceMovement = Date.now() - this.lastMovementTime;
      this.isMoving = timeSinceMovement < this.stoppedDuration;
    }
  }
  
  // Improved status update with continuous loop logic
  updateStatus() {
    const currentTime = Date.now();
    const timeSinceMovement = currentTime - this.lastMovementTime;
    
    if (timeSinceMovement >= this.stoppedDuration) {
      // Stopped for 30+ minutes
      this.currentStatus = 'Stopped';
    } else {
      // Running - show continuous loop heading status
      this.currentStatus = 'Running';
    }
    
    // Notify status change if needed
    this.checkStatusChange();
  }
  
  // Get heading status - always "Heading to" in continuous loop
  getHeadingStatus() {
    if (this.nextBuilding && this.currentStatus === 'Running') {
      return `Heading to ${this.nextBuilding.displayName}`;
    }
    return null;
  }
  
  // Check for status changes and notify subscribers
  checkStatusChange() {
    if (this.currentStatus !== this.lastNotifiedStatus) {
      console.log(`📊 Frontend Tram status: ${this.lastNotifiedStatus || 'Unknown'} → ${this.currentStatus}`);
      
      // Notify all status change callbacks
      this.statusChangeCallbacks.forEach(callback => {
        try {
          callback({
            oldStatus: this.lastNotifiedStatus,
            newStatus: this.currentStatus,
            location: this.currentLocation,
            headingTo: this.getHeadingStatus(),
            timestamp: Date.now()
          });
        } catch (error) {
          console.error('Error calling status change callback:', error);
        }
      });
      
      this.lastNotifiedStatus = this.currentStatus;
    }
  }
  
  // Subscribe to status changes
  onStatusChange(callback) {
    if (typeof callback === 'function') {
      this.statusChangeCallbacks.push(callback);
      return true;
    }
    return false;
  }
  
  // Unsubscribe from status changes
  offStatusChange(callback) {
    const index = this.statusChangeCallbacks.indexOf(callback);
    if (index > -1) {
      this.statusChangeCallbacks.splice(index, 1);
      return true;
    }
    return false;
  }
  
  // Calculate distance between two GPS coordinates in meters
  calculateDistance(lat1, lon1, lat2, lon2) {
    const R = 6371000; // Earth's radius in meters
    const dLat = this.degreesToRadians(lat2 - lat1);
    const dLon = this.degreesToRadians(lon2 - lon1);
    const a = Math.sin(dLat/2) * Math.sin(dLat/2) +
              Math.cos(this.degreesToRadians(lat1)) * Math.cos(this.degreesToRadians(lat2)) *
              Math.sin(dLon/2) * Math.sin(dLon/2);
    const c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1-a));
    return R * c;
  }
  
  degreesToRadians(degrees) {
    return degrees * (Math.PI/180);
  }
  
  // Improved tracking info with continuous loop
  getTrackingInfo() {
    return {
      status: this.currentStatus,
      headingTo: this.getHeadingStatus(),
      isMoving: this.isMoving,
      currentLocation: this.currentLocation,
      lastPassedBuilding: this.lastPassedBuilding,
      nextBuilding: this.nextBuilding,
      timestamp: Date.now(),
      locationHistoryLength: this.locationHistory.length,
      timeSinceLastMovement: Date.now() - this.lastMovementTime
    };
  }
  
  // Improved API status with continuous loop
  getStatusForAPI() {
    const info = this.getTrackingInfo();
    
    return {
      tram_id: 'tram_01_frontend',
      currentStatus: this.currentStatus,
      headingTo: info.headingTo,
      location: {
        lat: info.currentLocation?.lat || null,
        lng: info.currentLocation?.lon || null
      },
      last_building: info.lastPassedBuilding?.name || null,
      timestamp: info.timestamp,
      is_moving: this.isMoving,
      time_since_movement_ms: info.timeSinceLastMovement
    };
  }
  
  // Reset tracking state
  reset() {
    this.currentStatus = 'Stopped';
    this.currentLocation = null;
    this.lastLocation = null;
    this.isMoving = false;
    this.nextBuilding = null;
    this.lastPassedBuilding = null;
    this.locationHistory = [];
    this.lastUpdateTime = Date.now();
    this.lastMovementTime = Date.now();
  }
  
  // Get all buildings
  getBuildings() {
    return this.buildings;
  }
}

export default TramTracker; 

File: backend/tram_tracker.py
"""
Tram Tracking Logic for AU University Tram System
"""
import time
import math
import threading
from datetime import datetime
from gps_data import GPS_POINTS, BUILDINGS

class TramTracker:
    def __init__(self):
        self.current_position = None
        self.last_update_time = None
        self.last_movement_time = time.time()  # Track when tram last moved
        self.location_history = []
        self.max_history = 10
        
        # Route tracking - simplified for continuous loop
        self.last_passed_building = None
        self.next_building = None
        self.has_left_building_radius = True  # Track if tram has left the last building's radius
        
        # Movement detection settings
        self.movement_threshold = 0.00001  # GPS coordinate difference to detect movement
        self.stopped_duration = 30 * 60  # 30 minutes in seconds
        
        # Simulation
        self.simulation_thread = None
        self.simulation_running = False
        self.simulation_index = 0
        
        # Debug logging throttle
        self.last_debug_log_time = 0
        self.debug_log_interval = 5  # seconds
        
        # Initialize with MSM as starting point
        self.initialize_route()
        
    def initialize_route(self):
        """Initialize route with MSM as starting point"""
        self.last_passed_building = None
        self.next_building = BUILDINGS[0]  # MSM Building
        self.has_left_building_radius = True
        
    def get_next_building(self, current_building_id):
        """Get the next building in continuous loop: MSM → IT → AU Mall → Queen of Sheba → MSM"""
        # Define the continuous loop sequence
        loop_sequence = ["msm_building", "it_building", "au_mall", "queen_of_sheba"]
        
        if not current_building_id:
            return BUILDINGS[0]  # Start with MSM
            
        try:
            # Find current building in the sequence
            current_index = next(i for i, building_id in enumerate(loop_sequence) if building_id == current_building_id)
            # Get next building in loop (wraps around to MSM after Queen of Sheba)
            next_index = (current_index + 1) % len(loop_sequence)
            next_building_id = loop_sequence[next_index]
            
            # Find the building object
            return next(b for b in BUILDINGS if b["id"] == next_building_id)
            
        except StopIteration:
            return BUILDINGS[0]  # Default to MSM if not found
    
    def detect_movement(self, lat, lon):
        """Detect if the tram has moved significantly"""
        if not self.current_position:
            return True  # First position update counts as movement
            
        # Calculate distance from last position
        distance = self.calculate_distance(
            self.current_position["lat"], self.current_position["lon"],
            lat, lon
        )
        
        # Convert to meters (approximate)
        distance_meters = distance * 111000
        
        # If moved more than ~3 meters, consider it movement (more sensitive)
        if distance_meters > 3:
            self.last_movement_time = time.time()
            return True
            
        return False
    
    def update_position(self, lat, lon):
        """Update tram position and detect buildings"""
        current_time = time.time()
        
        # Detect movement
        has_moved = self.detect_movement(lat, lon)
        
        # Update current position
        self.current_position = {
            "lat": lat,
            "lon": lon,
            "timestamp": current_time
        }
        
        # Debug logging for position updates
        if has_moved:
            print(f"📍 Tram position updated: {lat:.6f}, {lon:.6f}")
        
        # Add to history
        self.location_history.append(self.current_position.copy())
        if len(self.location_history) > self.max_history:
            self.location_history.pop(0)
            
        # Check for nearby buildings
        detected_building = self.detect_building(lat, lon)
        
        if detected_building:
            # We're inside a building's radius
            if not self.last_passed_building or detected_building["id"] != self.last_passed_building["id"]:
                # New building detected
                print(f"🏢 Tram entered building: {detected_building['displayName']}")
                self.last_passed_building = detected_building
                self.next_building = self.get_next_building(detected_building["id"])
                self.has_left_building_radius = False
                print(f"🎯 Next destination updated: {self.next_building['displayName']}")
        else:
            # We're outside any building radius
            if self.last_passed_building and not self.has_left_building_radius:
                # We just left a building's radius
                print(f"🚪 Tram left building: {self.last_passed_building['displayName']}")
                self.has_left_building_radius = True
            
        self.last_update_time = current_time
        
        return self.get_tracking_info()
    
    def detect_building(self, lat, lon):
        """Detect if tram is near a building"""
        current_time = time.time()
        should_log = (current_time - self.last_debug_log_time) >= self.debug_log_interval
        
        for building in BUILDINGS:
            distance = self.calculate_distance(lat, lon, building["lat"], building["lon"])
            # Debug: print distance to each building (throttled)
            if should_log:
                distance_meters = distance * 111000
                print(f"🏢 Distance to {building['displayName']}: {distance_meters:.1f}m (radius: {building['radius']*111000:.1f}m)")
            if distance <= building["radius"]:
                return building
        
        if should_log:
            self.last_debug_log_time = current_time
        
        return None
    
    def calculate_distance(self, lat1, lon1, lat2, lon2):
        """Calculate distance between two GPS coordinates in degrees"""
        return math.sqrt((lat1 - lat2) ** 2 + (lon1 - lon2) ** 2)
    
    def is_tram_stopped(self):
        """Check if tram has been stopped for 30+ minutes"""
        current_time = time.time()
        time_since_movement = current_time - self.last_movement_time
        return time_since_movement >= self.stopped_duration
    
    def get_status(self):
        """Get current tram status"""
        # Check if stopped for 30+ minutes
        if self.is_tram_stopped():
            status = "Stopped"
        else:
            status = "Running"
        
        # Always show "Heading to" next building when tram is running (continuous loop)
        if self.next_building and not self.is_tram_stopped():
            heading_to = f"Heading to {self.next_building['displayName']}"
        else:
            heading_to = None
            
        return {
            "status": status,
            "headingTo": heading_to,
            "lastPassedBuilding": self.last_passed_building,
            "nextBuilding": self.next_building,
            "currentLocation": self.current_position,
            "timestamp": int(time.time() * 1000),
            "timeSinceMovement": int(time.time() - self.last_movement_time)
        }
    
    def get_tracking_info(self):
        """Get detailed tracking information"""
        status_info = self.get_status()
        
        return {
            "status": status_info["status"],
            "headingTo": status_info["headingTo"],
            "lastPassedBuilding": status_info["lastPassedBuilding"],
            "nextBuilding": status_info["nextBuilding"],
            "currentLocation": status_info["currentLocation"],
            "locationHistoryLength": len(self.location_history),
            "timestamp": status_info["timestamp"],
            "timeSinceMovement": status_info["timeSinceMovement"]
        }
    
    def get_status_for_api(self):
        """Get status formatted for iOS API"""
        status = self.get_status()
        
        # Format heading text - always "Heading to" in continuous loop
        heading_text = None
        if status["headingTo"]:
            heading_text = f"Heading to {status['nextBuilding']['displayName']}"
        
        return {
            "tram_id": "tram_01",
            "currentStatus": status["status"],
            "headingTo": heading_text,
            "location": {
                "lat": self.current_position["lat"] if self.current_position else None,
                "lng": self.current_position["lon"] if self.current_position else None
            },
            "last_building": status["lastPassedBuilding"]["name"] if status["lastPassedBuilding"] else None,
            "timestamp": status["timestamp"],
            "time_since_movement_seconds": status["timeSinceMovement"]
        }
    
    def start_simulation(self, speed_multiplier=1.0):
        """Start tram simulation along the route"""
        if self.simulation_running:
            return {"message": "Simulation already running"}
            
        self.simulation_running = True
        
        # Reset movement time when starting simulation
        self.last_movement_time = time.time()
        
        self.simulation_thread = threading.Thread(
            target=self._simulation_loop,
            args=(speed_multiplier,)
        )
        self.simulation_thread.daemon = True
        self.simulation_thread.start()
        
        return {
            "message": "Simulation started",
            "speed_multiplier": speed_multiplier,
            "total_points": len(GPS_POINTS)
        }
    
    def stop_simulation(self):
        """Stop tram simulation"""
        self.simulation_running = False
        if self.simulation_thread:
            self.simulation_thread.join(timeout=1)
        return {"message": "Simulation stopped"}
    
    def _simulation_loop(self, speed_multiplier):
        """Simulation loop that moves tram along GPS points"""
        while self.simulation_running:
            # Get current GPS point
            gps_point = GPS_POINTS[self.simulation_index]
            
            # Update position
            self.update_position(gps_point["lat"], gps_point["lon"])
            
            # Move to next point
            self.simulation_index = (self.simulation_index + 1) % len(GPS_POINTS)
            
            # Sleep based on speed multiplier
            time.sleep(0.5 / speed_multiplier)
    
    def reset(self):
        """Reset tram tracker to initial state"""
        self.current_position = None
        self.last_update_time = None
        self.last_movement_time = time.time()
        self.location_history = []
        self.has_left_building_radius = True
        self.initialize_route()
        self.stop_simulation()
        self.simulation_index = 0 